core/src/main/scala/dagr/core/execsystem/TaskTracker.scala

```diff
@@ -0,0 +1,344 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2015 Fulcrum Genomics LLC
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package dagr.core.execsystem
+
+import java.nio.file.Path
+import java.sql.Timestamp
+
+import dagr.DagrDef._
+import dagr.core.execsystem.TaskStatus._
+import dagr.core.execsystem.TaskManagerLike.BaseInfoAndNode
+import dagr.core.execsystem.TaskTracker.InfoAndNode
+import dagr.core.tasksystem.Task
+import dagr.core.util.{BiMap, LazyLogging}
+
+import scala.collection.mutable
+
+object TaskTracker {
+  case class InfoAndNode(info: TaskExecutionInfo, node: GraphNode) extends BaseInfoAndNode(info=info, node=node)
+}
+
+/** Tracks tasks during their execution */
+trait TaskTracker extends TaskManagerLike with LazyLogging {
+
+  /** Stores information about the execution of the current task and its place in the execution graph.  Mainly used
+    * for convenience during lookups.
+    */
+  private final class TrackingInfo(task: Task, parent: Option[GraphNode] = None) {
+    val (info: TaskExecutionInfo, node: GraphNode) = {
+      val id: BigInt = task.id.getOrElse(unreachable(s"Task id not found for task with name '${task.name}'"))
+      val submissionDate: Option[Timestamp] = Some(new Timestamp(System.currentTimeMillis))
+      val info = new TaskExecutionInfo(
+        task=task,
+        status=UNKNOWN,
+        script=scriptPathFor(task=task, id=id),
+        logFile=logPathFor(task=task, id=id),
+        submissionDate=submissionDate
+      )
+      val node = predecessorsOf(task = task) match {
+        case None => new GraphNode(task=task, predecessorNodes=Nil, state=GraphNodeState.ORPHAN, parent=parent)
+        case Some(predecessors) => new GraphNode(task=task, predecessorNodes=predecessors, parent=parent)
+      }
+      (info, node)
+    }
+  }
+
+  /** The next id for a task */
+  private var nextId: BigInt = 0
```

Comment:
Also, since it's used in a lot of places it might make sense to have a `type TaskId = BigInt`, which would make it easy to change in the future, and also clearer what it is everywhere.